alerts/apps_alerts.libsonnet

```diff
@@ -264,6 +264,35 @@
               summary: 'Job failed to complete.',
             },
           },
+          {
+            alert: 'KubeJobFailing',
+            expr: |||
+              clamp_max(
+                max(
+                  kube_job_status_start_time{%(prefixedNamespaceSelector)s%(kubeStateMetricsSelector)s}
+                  * ON(job_name,namespace) GROUP_RIGHT()
+                  kube_job_owner{owner_name!="",owner_name!="<none>"}
+                )
+                BY (job_name, owner_name, namespace)
+                == ON(owner_name) GROUP_LEFT()
+                max(
+                  kube_job_status_start_time{%(prefixedNamespaceSelector)s%(kubeStateMetricsSelector)s}
+                  * ON(job_name,namespace) GROUP_RIGHT()
+                  kube_job_owner{owner_name!="",owner_name!="<none>"}
+                )
+              BY (owner_name), 1)
+              * ON(job_name) GROUP_LEFT()
+              (kube_job_status_failed{%(prefixedNamespaceSelector)s%(kubeStateMetricsSelector)s} > 0)
```

Comment:
I am worried about the complexity of this query. Did you do some evaluation on the impact of the rules engine and the amount of data processed here?

Reply:
I will check how this could be evaluated. I could rewrite the query into two recording rules with alerting rule looking similar to the query below:
```
clamp_max(job:kube_job_status_start_time:max, 1)
* ON(job_name) GROUP_LEFT()
(kube_job_status_failed{%(prefixedNamespaceSelector)s%(kubeStateMetricsSelector)s} > 0)
```

Reply:
Bellow is performance from [query log](https://prometheus.io/docs/guides/query-log/) of KubeJob**Failing**
```
"stats": {
    "timings": {
      "evalTotalTime": 0.002736879,
      "resultSortTime": 0,
      "queryPreparationTime": 0.001494211,
      "innerEvalTime": 0.001230275,
      "execQueueTime": 0.000011588,
      "execTotalTime": 0.002754576
    },
    "samples": {
      "totalQueryableSamples": 182,
      "peakSamples": 206
    }
  },
```
vs KubeJob**Failed**
```
"stats": {
    "timings": {
      "evalTotalTime": 0.000277071,
      "resultSortTime": 0,
      "queryPreparationTime": 0.000185315,
      "innerEvalTime": 0.000087251,
      "execQueueTime": 0.000003611,
      "execTotalTime": 0.00028435
    },
    "samples": {
      "totalQueryableSamples": 24,
      "peakSamples": 58
    }
  },
```

Reply:
I will compare it with same query but split in recording rules as mentioned earlier.

Reply:
Performance of KubeJob**Failing** with use of recording rules
```
  "stats": {
    "timings": {
      "evalTotalTime": 0.000071755,
      "resultSortTime": 0,
      "queryPreparationTime": 0.000040308,
      "innerEvalTime": 0.000022829,
      "execQueueTime": 0.000007586,
      "execTotalTime": 0.00008497
    },
    "samples": {
      "totalQueryableSamples": 1,
      "peakSamples": 1
    }
  },
```
@s-urbaniak from the impact of the rules engine and the amount of data processed point of view this would be fine, right? Alerting rule would contain only `job:kube_job_status_failing:sum` with rest covered by recording rules.